src/main/PersistentState.cpp

```diff
@@ -4,32 +4,111 @@
 #include "PersistentState.h"
 
+#include "crypto/Hex.h"
 #include "database/Database.h"
 #include "herder/Herder.h"
+#include "herder/HerderUtils.h"
 #include "ledger/LedgerManager.h"
 #include "util/GlobalChecks.h"
 #include "util/Logging.h"
-#include "xdr/Stellar-internal.h"
 #include <Tracy.hpp>
 
 namespace stellar
 {
 
 using namespace std;
 
+std::chrono::minutes const PersistentState::TX_SET_GC_DELAY(1);
+
 std::string PersistentState::mapping[kLastEntry] = {
     "lastclosedledger", "historyarchivestate", "lastscpdata",
     "databaseschema",   "networkpassphrase",   "ledgerupgrades",
-    "rebuildledger",    "lastscpdataxdr"};
+    "rebuildledger",    "lastscpdataxdr",      "txset"};
 
 std::string PersistentState::kSQLCreateStatement =
     "CREATE TABLE IF NOT EXISTS storestate ("
     "statename   CHARACTER(32) PRIMARY KEY,"
     "state       TEXT"
     "); ";
 
-PersistentState::PersistentState(Application& app) : mApp(app)
+PersistentState::PersistentState(Application& app)
+    : mApp(app), mTxSetGarbageCollectTimer(app)
+{
+}
+
+void
+PersistentState::startTxSetGCTimer()
+{
+    mTxSetGarbageCollectTimer.expires_from_now(TX_SET_GC_DELAY);
+    mTxSetGarbageCollectTimer.async_wait(
+        [this]() {
+            try
+            {
+                this->deleteOldTxSets();
+            }
+            catch (std::exception& e)
+            {
+                CLOG_ERROR(Herder, "Error while deleting old tx sets: {}",
+                           e.what());
+            }
+        },
+        &VirtualTimer::onFailureNoop);
+}
+
+void
+PersistentState::shutdown()
 {
+    mTxSetGarbageCollectTimer.cancel();
+}
+
+void
+PersistentState::deleteOldTxSets()
+{
+    auto hashesToDelete = getTxSetHashesForAllSlots();
+
+    for (auto const& state : getSCPStateAllSlots())
+    {
+        std::vector<uint8_t> buffer;
+        decoder::decode_b64(state, buffer);
+        PersistedSCPState scpState;
+        try
+        {
+            xdr::xdr_from_opaque(buffer, scpState);
+            for (auto const& e : scpState.v1().scpEnvelopes)
+            {
+                for (auto const& hash : getTxSetHashes(e))
+                {
+                    hashesToDelete.erase(hash);
+                }
+            }
+        }
+        catch (std::exception& e)
+        {
+            CLOG_ERROR(Herder, "Error while deleting old tx sets: {}",
+                       e.what());
+        }
+    }
+
+    for (auto const& hash : hashesToDelete)
+    {
+        auto name = getStoreStateNameForTxSet(hash);
+        auto prep = mApp.getDatabase().getPreparedStatement(
+            "DELETE FROM storestate WHERE statename = :n;");
+
+        auto& st = prep.statement();
+        st.exchange(soci::use(name));
+        st.define_and_bind();
+        st.execute(true);
+    }
+
+    startTxSetGCTimer();
+}
+
+void
+PersistentState::upgradeSizeLimit(Database& db)
+{
+    db.getSession()
+        << "ALTER TABLE storestate ALTER COLUMN statename TYPE CHARACTER(70);";
```

Comment:
why 70? Is it just max string size `PersistentState::mapping` + 32 (to store hash) + 1 (for extra characters) ? It seems quite close to what we do (this max size does not really do anything, so may break again in the future)